secure/storage/src/crypto_kv_storage.rs

```diff
@@ -0,0 +1,177 @@
+            _ => Err(Error::UnexpectedValueType),
+        }
+    }
+
+    fn get_private_key_for_name_and_version(
+        &self,
+        key_pair_name: &str,
+        key_pair_version: Ed25519PublicKey,
+    ) -> Result<Ed25519PrivateKey, Error> {
+        let current_private_key = self.get_private_key_for_name(key_pair_name)?;
+        if current_private_key.public_key().eq(&key_pair_version) {
+            return Ok(current_private_key);
+        }
+
+        let previous_private_key =
+            self.get_private_key_for_name(&Self::get_previous_version_name(key_pair_name))?;
+        if previous_private_key.public_key().eq(&key_pair_version) {
+            return Ok(previous_private_key);
+        }
+
+        Err(Error::KeyVersionNotFound(key_pair_version.to_string()))
+    }
+
+    fn rotate_key_pair(&mut self, key_pair_name: &str) -> Result<Ed25519PublicKey, Error> {
+        match self.storage.get(key_pair_name)? {
+            Value::Ed25519PrivateKey(private_key) => {
+                let (new_private_key, new_public_key) = Self::new_ed25519_key_pair()?;
+                self.storage.set(
+                    &Self::get_previous_version_name(key_pair_name),
+                    Value::Ed25519PrivateKey(private_key),
+                )?;
+                self.storage
+                    .set(key_pair_name, Value::Ed25519PrivateKey(new_private_key))?;
+                Ok(new_public_key)
+            }
+            _ => Err(Error::UnexpectedValueType),
+        }
+    }
+
+    fn sign_message(
+        &mut self,
+        key_pair_name: &str,
+        message: &HashValue,
+    ) -> Result<Ed25519Signature, Error> {
+        let private_key = self.get_private_key_for_name(key_pair_name)?;
+        Ok(private_key.sign_message(message))
+    }
+
+    fn sign_message_using_version(
+        &mut self,
+        key_pair_name: &str,
+        key_pair_version: Ed25519PublicKey,
+        message: &HashValue,
+    ) -> Result<Ed25519Signature, Error> {
+        let private_key =
+            self.get_private_key_for_name_and_version(key_pair_name, key_pair_version)?;
+        Ok(private_key.sign_message(message))
+    }
+}
+
+/// Delegates the KVStorage implementation to the internal storage engine.
+impl KVStorage for CryptoKVStorage {
+    fn available(&self) -> bool {
+        self.storage.available()
+    }
+
+    fn create(&mut self, key: &str, value: Value, policy: &Policy) -> Result<(), Error> {
+        self.storage.create(key, value, policy)
+    }
+
+    fn get(&self, key: &str) -> Result<Value, Error> {
+        self.storage.get(key)
+    }
+
+    fn set(&mut self, key: &str, value: Value) -> Result<(), Error> {
+        self.storage.set(key, value)
+    }
+
+    #[cfg(test)]
+    fn reset_and_clear(&mut self) -> Result<(), Error> {
+        self.storage.reset_and_clear()
+    }
+}
+
+impl Storage for CryptoKVStorage {}
+
+impl CryptoKVStorage {
+    pub fn new(storage: Box<dyn KVStorage>) -> Self {
```

Comment:
This should ideally not be a generic not a box.

Reply:
Thanks. I've changed it to a generic. Although, one thing to note: it now means the clients must specify the type (e.g., when getting test implementations of CryptoKVStorage).

Reply:
But shouldn't this be hidden by the implementation as OnDisk would specify CryptoKVStorage<OnDisk>?

Reply:
Thanks, David! As discussed offline, I've hidden the generics by making CryptoKVStorage extend KVStorage, giving CryptoKVStorage a default implementation to implement CryptoStorage, and allowing CryptoKVStorage to implement Storage. Now the convenience functions doesn't expose generics.